### ISSUE TITLE
Is linear indexing a matrix well-defined?

### ISSUE THREAD

**Issue by cbm755** [sympy/sympy#24007]
Suppose `M` is a Matrix (Mutable or Immutable).  Is it intended that `M[0]` is the same as `M[0, 0]`?.  That is loop over linear indexing of a Matrix visits the entries in some order (column-major, row-major is not important for the question).

Example:
```
>>> M = Matrix(2, 3, range(10, 16))
>>> M[0]
10
>>> M[0,0]
10
>>> M[1]
11
>>> M[0, 1]
11
```

I found no mention of this [in the docs](https://docs.sympy.org/latest/modules/matrices/matrices.html) so I suppose it might be unsupported (?)

### background

I notice that `Array` does NOT behave like this (while porting code viz https://docs.sympy.org/latest/explanation/active-deprecations.html#deprecated-non-expr-in-matrix), instead it gives you the 0th row, the 1st row etc.  I'm wondering which behaviour is intended... 

### going forward

If not intended, maybe we should raise exception on linear index of Matrix?

**Comment by cbm755** [comment]
Ok, so it is documented so I guess my question becomes: is it true that `Array` is NOT supposed to linear index in this fashion?   Perhaps I can read and answer this myself too... ;-)

**Comment by oscarbenjamin** [comment]
I think the idea is that Array should be more consistent with numpy arrays but honestly I'm not sure because multiplication of arrays is disallowed. Anyway numpy arrays index in such a way that you always drop 1 dimension:
```python
In [24]: z = np.zeros((3, 4, 5))

In [25]: z.shape
Out[25]: (3, 4, 5)

In [26]: z[0].shape
Out[26]: (4, 5)

In [27]: z[0][0].shape
Out[27]: (5,)

In [28]: z[0][0][0].shape
Out[28]: ()
```
That makes sense in an n-dimensional context. With matrices that are always 2D it doesn't make sense so there are two obvious possibilities:

1. Give 1xn matrices for the rows.
2. Give the elements of the matrix in some order.

You can see that numpy's matrix goes for the first of these:
```python
In [29]: m = np.matrix([[1, 2, 3], [4, 5, 6]])

In [30]: m.shape
Out[30]: (2, 3)

In [31]: m[0].shape
Out[31]: (1, 3)

In [32]: m[0][0].shape
Out[32]: (1, 3)

In [33]: m[0][0][0].shape
Out[33]: (1, 3)
```
This gives it the weird infinitely indexable-ness that strings also have when it isn't possible to distinguish a sequence of length 1 from the element in that sequence.

SymPy's matrix went with the other option to index the elements. There can be some cases where that's useful because in some contexts you might just want to iterate over all of the elements without caring where they are in the matrix. Really though I think that because the old matrix class stored all of the elements in a flat list it was just something that could be conveniently added and there hasn't been a strong reason to change that.

Personally I think that the linear indexing was a mistake and it would be better just to have a method that gives a flat array or a list of the elements e.g.:
```python
In [40]: M = Matrix(3, 2, range(6))

In [41]: M.flat()
Out[41]: [0, 1, 2, 3, 4, 5]
```
That's also how numpy provides it:
```python
In [42]: z.flat
Out[42]: <numpy.flatiter at 0x3f99850>

In [43]: z.flatten()
Out[43]: 
array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0.])
```

**Comment by cbm755** [comment]
I think I agree.  Maybe a round of deprecation for linear indexing into Matrix is a good idea.

**Comment by cbm755** [comment]
Not sure this is relevant but somewhat related is the behaviour of `len` on Array:

```
A = Array(range(10,16), shape=(2,3))
⎡10  11  12⎤
⎢          ⎥
⎣13  14  15⎦

>>> len(A)
6

>>> len([x for x in A])
2
```

I checked numpy and it **does not** behave like that.  I guess that should be a separate issue...

```
>>> z = np.zeros((10, 11, 12))
>>> len(z)
10

>>> len([x for x in z])
10
```

**Comment by oscarbenjamin** [comment]
> Maybe a round of deprecation for linear indexing into Matrix is a good idea.

I do worry about all this deprecating. It seemed very clear to me that the elements of a Matrix should be Expr and that attempting to accommodate for anything else was an unreasonable restriction preventing improvements to the Matrix code for the important common use cases. This isn't the first time I've seen that downstream code was depending on putting arbitrary objects in there though.

**Comment by asmeurer** [comment]
I also agree that the linear indexing was a bad idea, and we should be consistent with NumPy. This was discussed here too https://github.com/sympy/sympy/issues/20410#issuecomment-725916905. IMO indexing into a matrix should work just like NumPy except that the result is always 2-D (i.e., a Matrix).

Funnily enough, one of the NumPy developers actually likes linear indexing and wishes NumPy did that (https://github.com/data-apis/array-api/issues/188). 

`len()` is closely related. Generally speaking, if an object in Python is indexable, it should satisfy `list(iter(obj)) == [obj[i] for i in range(len(obj))]`. So in that sense, the linear indexing is nice because it means `len` gives the total size of the matrix instead of just the number of rows. An option is also to make `__len__` not be implemented, and make people use `A.shape` or `A.rows` (and we could add `Matrix.size` to correspond to `np.ndarray.size`).

> I think the idea is that Array should be more consistent with numpy arrays but honestly I'm not sure because multiplication of arrays is disallowed

Indexing and vectorized operations are separate things. Indexing an Array should work just like in NumPy, because NumPy has pretty good indexing semantics. Whether we want to support other NumPy array features like vectorization on operations and broadcasting is a separate discussion. Personally, I think they should be supported, but in a more explicit way (there should be explicit expression tree nodes for VectorizedOp or Broadcasted). That's cleaner for code generation, especially to languages where those are more explicit anyway.

**Comment by asmeurer** [comment]
I'm also not sure if it's actually worth changing the way `Matrix` indexing works. It's been that way for a long time. `Matrix` is always 2-D so the expressiveness of multidimensional indexing is inherently limited anyway. But I definitely think `Array` indexing should work like NumPy and if there are any discrepancies there, they should be fixed.

**Comment by asmeurer** [comment]
Also a good idea to document all this a little better, especially if we decide to keep things the way they are.

**Comment by 1e9abhi1e10** [comment]
Hello everyone, I wish to fix this issue could you please assign this issue to me?
